srml/support/src/traits.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Traits for SRML
+
+use crate::rstd::result;
+use crate::codec::Codec;
+use crate::runtime_primitives::traits::{MaybeSerializeDebug, SimpleArithmetic, As};
+
+/// The account with the given id was killed.
+pub trait OnFreeBalanceZero<AccountId> {
+	/// The account was the given id was killed.
+	fn on_free_balance_zero(who: &AccountId);
+}
+
+impl<AccountId> OnFreeBalanceZero<AccountId> for () {
+	fn on_free_balance_zero(_who: &AccountId) {}
+}
+impl<
+	AccountId,
+	X: OnFreeBalanceZero<AccountId>,
+	Y: OnFreeBalanceZero<AccountId>,
+> OnFreeBalanceZero<AccountId> for (X, Y) {
+	fn on_free_balance_zero(who: &AccountId) {
+		X::on_free_balance_zero(who);
+		Y::on_free_balance_zero(who);
+	}
+}
+
+/// Trait for a hook to get called when some balance has been minted, causing dilution.
+pub trait OnDilution<Balance> {
+	/// Some `portion` of the total balance just "grew" by `minted`. `portion` is the pre-growth
+	/// amount (it doesn't take account of the recent growth).
+	fn on_dilution(minted: Balance, portion: Balance);
+}
+
+impl<Balance> OnDilution<Balance> for () {
+	fn on_dilution(_minted: Balance, _portion: Balance) {}
+}
+
+/// Determinator for whether a given account is able to transfer balance.
+pub trait EnsureAccountLiquid<AccountId> {
+	/// Returns `Ok` iff the account is able to transfer funds normally. `Err(...)`
+	/// with the reason why not otherwise.
+	fn ensure_account_liquid(who: &AccountId) -> result::Result<(), &'static str>;
+}
+impl<
+	AccountId,
+	X: EnsureAccountLiquid<AccountId>,
+	Y: EnsureAccountLiquid<AccountId>,
+> EnsureAccountLiquid<AccountId> for (X, Y) {
+	fn ensure_account_liquid(who: &AccountId) -> result::Result<(), &'static str> {
+		X::ensure_account_liquid(who)?;
+		Y::ensure_account_liquid(who)
+	}
+}
+impl<AccountId> EnsureAccountLiquid<AccountId> for () {
+	fn ensure_account_liquid(_who: &AccountId) -> result::Result<(), &'static str> { Ok(()) }
+}
+
+/// Outcome of a balance update.
+pub enum UpdateBalanceOutcome {
+	/// Account balance was simply updated.
+	Updated,
+	/// The update has led to killing of the account.
+	AccountKilled,
+}
+
+/// Abstraction over staking mecanism
```

Comment:
This is not an abstraction over the staking mechanism - it's an abstraction over a fungible assets system.